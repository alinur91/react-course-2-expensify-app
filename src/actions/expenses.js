import uuid from 'uuid';

// ADD_EXPENSE
export const addExpense = (
    {
        description='',note='', /* default values */
        amount=0,
        createdAt=0
    }={}
) => ({ 
/* if object doesnt exist we will desctrutre empty object */
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(),
        description,
        note,
        amount,
        createdAt
    }
})

//REMOVE_EXPENSE
export const removeExpense = ({id}={}) =>( /* {id}={} degen zna4it dispatch(removeExpense({id})) degen soz object peredaem */
    {
        type: 'REMOVE_EXPENSE',
        id
    }
)

//EDIT_EXPENSE
export const editExpense = (id,updates) =>(
    {
        type: 'EDIT_EXPENSE',
        id,
        updates
    }
)