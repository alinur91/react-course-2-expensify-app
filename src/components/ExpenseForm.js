import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates'; /* SingleDatePicker degen component */
/* SingleDatePicker component do have required props */
//const date = new Date()
// const now = moment() /*we get back instance of moment,if u don't provide anything its now  */
// console.log(now.format('MMM Do, YYYY')) /* format() method its going to print formatted version of the now */


export default class ExpenseForm extends React.Component {
/*The goal is to take these 4 state values and only use  defaults if no expense passed down  */
/* if an expense was passed down we want to start these off that those values(props.expense) */
/*To get that done we have to look at props.We have to define our state in the constructor function to access the props  */
//props.expense.createdAt timeStamp.<ExpenseForm/> componenty we reuse v AddExpensePage i EditExpensePage componentax poetomy kogda v <AddExpensePage/> componente use descrip,amount bari empty boladi,kogda v <EditExpensePage/>-e ispolzuem ExpenseForm dannyie dolzny byt populated
constructor(props){ /* state vklu4ili v constr function shtoby byl access to props() */
    super(props)
    this.state = {
        description: props.expense?props.expense.description:'',
        note: props.expense?props.expense.note:'',
        amount: props.expense?(props.expense.amount / 100).toString():'',
        createdAt: props.expense?moment(props.expense.createdAt):moment(),
        calenderFocused: false,
        error: ''
    }
}
    
    onDescriptionChange=e =>{
  
        const description= e.target.value /* inputtin valuei descrip: 'asd' */
        this.setState(()=>({description})) /*img gonna pass in updated function,will return implicitly object  */
    }
    onNoteChange = e=>{ /* kazhdyi raz kak e.target.value changes v object {note: e.target.value} budet i textarea nadpis budet 4itat s value={this.state.note} statea */
        const note =e.target.value
        // e.persist()
        this.setState(()=>({note}))
    }
    onAmountChange = e =>{
        /* we're not going to blindly set,conditional logic bolady */
        const amount = e.target.value; /*e.target.value; degen svezhyi value nado setStatepen v objecte zapisat  */
//!amount degen empty string bolsa(falsy) vse ravno input value set it isteimiz
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){this.setState(()=>({amount}))}
        /* tobedegi crietria shak kelmese we're not going to setState */
    }/* amount.match(/^\d*(\.\d{0,2})?$/) degen regular expression 12.12 boly kerek nukteden keyin 2 cifra kerek */
/* the big picture goal here is to use local component state to track the changes to all of these inputs,only when the user submits the form will do smth with that information */
/* When they submit the from will send that off to redux to either edit the existing expense or create the new one*/
/* value={this.state.description}} degen creates a read only input,we want to user to be able to change the value */
onDateChange = createdAt =>{ /* kogda user menyaet daty,on menyaetsya na to shto on postavil,createdAt degen shto on postavil */
/*onDateChange function boladi,its got to get called with moment() instance when someone picks a new day from the calendar  */
if(createdAt){this.setState(()=>({createdAt}))}
    /* if there is no createdAt(esli udalit delete keymen) do not nothing */
/* this is going to prevent the user from being able to clear that value */
}
onFocusChange= ({focused}) =>{
    
    this.setState(()=>({calenderFocused:focused}))
}
onSubmit = e =>{
    e.preventDefault() /* full page refesh bolmaidy kogda button baskan kezi */
/* description men amount required field,calendarmen,note optional */
    if(!this.state.description || !this.state.amount){ /* on uzhe napisal description i amount field,kogda submit nazhimaet proveryaem esli ili net descr,amount,esli net 1-o iz dvux */
/* descr ili amount zhazbasa,hot' 1-ein,we want to render error to the screen */
        // Set error state equal to ''
        this.setState(()=>({error: 'Please provide description and amount.'}))

    }else{ /* if there is no error */
        // esli 1-yi raz s errorom bil mi hotim za4istit error
         this.setState(()=>({error: ''})) 
         this.props.onSubmit({ /* zdes peredaem object,v AddExpensePage prinimaem object dispatchim shtoby renderilsya expense,i redirecr props.history.push('/') */
             description: this.state.description,
             amount: parseFloat(this.state.amount) * 100, /*this.state.note degen 12.12 nado preobrazavat  */
             note: this.state.note,
             createdAt: this.state.createdAt.valueOf(), /*this.state.createdAt degen moment object  */
            
         }) /*this.props degen 4erez instance props onSubmit peredali <ExpenseForm  onSumbit(expense=>{console.log(expense)})/>onSubmit degen callback koi 4erep props peredaem i sdes vizivaem a v propse v functione polu4im expense,shtoby store.dispatch(addExpense(expense)) sdelat 4toby za renderit etot expense  */
    }
}
    render() { /*let's setup onSubmit handler for the form,kogda submit sdelaet redirect boladi dashboard page-ge  */
        return ( /* kogda addExpense nazhimaet button vizivaetsya onSubmit function i peredaet object a v EditExpesePage-e prinimaem onSubmit(expense)=>{} */
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input 
                        value={this.state.description} 
                        autoFocus 
                        placeholder="Description"
                        onChange={this.onDescriptionChange}
                    /> {/* autoFocus is going to make sure that when we visit the page it automatically puts the cursor and focus right in input */}
                    <input 
                        type="text" 
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calenderFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={(day)=> false} 
                     />
                    <textarea 
                        value={this.state.note}  
                        onChange={this.onNoteChange}
                        placeholder="Add a note for your expense (optional)">
                    </textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        )
    }/*date degen-moment object that represents where you want to start,we wanna start at now,when the user loads the page,bugingi kundi korsetemiz  */
//numberOfMonths={1} degen calendarda 1 ai korsetedi.isOutsideRange={(day)=> false}  degen otken shakti calendarda korsetedi
/* onFocusChange degen handler when react dates library needs to change that value */
}


/*How to simulate user interaction in your test cases.If I submit the form with bad data I would expect that error gets set and rendered to the screen. */
//We will write test case that makes sure an error gets rendered when the form is submitted without a description/an amount