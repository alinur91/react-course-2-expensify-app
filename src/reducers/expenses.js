// Expenses Reducer
const expensesReducerDefaultState = []

export default  (state=expensesReducerDefaultState ,action) =>{
    // console.log(state)
/* return a new state? */
switch (action.type) { /* push changes original array,we don't want to change state or action,we just want to read off of it */
case 'ADD_EXPENSE': return [...state,action.expense]//state degen: [{id:123,description:'asd',note:'',amount:123,createdAt: 456}] action.expense degen {id:123,description:'asd',note:'',amount:123,createdAt: 456} /* returns a new array,return array soxranitsya k expensesReducer variablu */
/* aside from concat we can use spread operator.action degen object vnutri property expense,id,amount */
    case 'REMOVE_EXPENSE': return state.filter(({id})=> id !==action.id)
    case 'EDIT_EXPENSE': return state.map(expense=>{
/* action degen   {type: 'EDIT_EXPENSE',id,updates} updates degen {amount: 900} */
/* state degen [{id: "bed68040-5eed-4835-81d9-1ec45e9f06bf", description: "Rent", note: "rent for may", amount: 800, createdAt},{id: "9f0a4e1c-1f5c-4085-a425-277adb9116e8", description: "Buy a car", note: "", amount: 10000, createdAt: 0}] */
        if(expense.id === action.id){
            return { /* i want to return a new object,changing expense,i want to grab all existing properties,and overwrite any of the ones that were passed down */
                ...expense,
                ...action.updates
            }
        }else{
            return expense
        }
    })
    default: return state    
}
}

