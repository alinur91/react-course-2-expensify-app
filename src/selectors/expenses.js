import moment from 'moment';

// Get visible expenses
export default (expenses,{text,sortBy,startDate,endDate}) => { /* vse dispatch sdelali neskolko zdes polnaya informacia,s pomoshui etogo vernem array [{},{}] */
    return expenses.filter(expense=>{
const createdAtMoment = moment(expense.createdAt) //expense.createdAt cifra goi
const startDateMatch= startDate? startDate.isSameOrBefore(createdAtMoment,'day') : true; //if there is no start date we're never going to filter things out based off of that,startDate always be true if there is no startdate
//typeof startDate !== 'number' || expense.createdAt >= startDate; /*only a start date is a number do we want to filter expenses out  */
/*lets say startdate is undefined,and createdAt is 1,it will stay(wont be filtered).Lets say startDate is 2 and created_at is 1(it will not stay,sozdannyi data ranshe bil,a user pozhe pokazivaet startDate) etot false budet startDateMatch  */
const endDateMatch= endDate? endDate.isSameOrAfter(createdAtMoment,'day') : true;
//typeof endDate !== 'number' || expense.createdAt <= endDate;
const textMatch=expense.description.toLowerCase().includes(text.toLowerCase())

        
/* esli vse eto startDateMatch && endDateMatch && textMatch bydet true to  expense ostaetsya v arraye*/
        return startDateMatch && endDateMatch && textMatch
    }).sort((a,b)=>{ /* a,b degen object {id: 123,description:'asd',createdAt: 123,amount:123} */
        if(sortBy === 'date') { /* sortiruem po recent data */
            return a.createdAt > b.createdAt?-1:1
        } // I want to see my most expenses coming 1st,from sorting date i want to see recent expenses coming 1st
        else if(sortBy === 'amount'){
            return b.amount - a.amount /* [1(a),2(b)] b-a = polozhitelnoe chislo,zna4it menyaem mestami,sonda kemy retimen boladi 800,100,50 */
           // a.amount > b.amount?-1:1
        }
    })
}

//JEST testing framework