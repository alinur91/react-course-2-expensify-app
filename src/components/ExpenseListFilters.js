//Filter by text value.We have no way to actually write to the store
/* We're going to learn how we can dispatch from a connected component */

import React from 'react'; /*import React from 'react'; vot eto delaem shtoby Componenty sozdavat  */
import {connect} from 'react-redux'; //zdez import delaem shtoby <ExpenseListFilters/> componenty podklu4it k store 4erez connect
import {DateRangePicker} from 'react-dates';
import {setTextFilter,sortByAmount,sortByDate,setStartDate,setEndDate} from '../actions/filters'; /* eto importiruem shtoby mozno bilo v componente props.dispatch(setTextFilter()) sdelat */
/* The goal from here is to get the old value off of the store*/
/* We set it to bill then water.Its really important that the input always matches up the current text value on the redux store.  store.dispatch(setTextFilter('bil')),store.dispatch(setTextFilter('water')) */
/* We're going to connect ExpenseListFilters to the store */
export class ExpenseListFilters extends React.Component {
    state = { /*  */
        calenderFocused: null
    }
    onDatesChange = ({startDate,endDate}) =>{ 
/* this function is going to get called with React dates library.Its going to get called with an object,on that obj we have startDate,endDate */
 /* we need to dispatch the correct actions to get the filters to change */       
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }

    onFocusChange= (calenderFocused) =>{
        this.setState(()=>({calenderFocused}))
    }

    onTextChange = e=>{
        this.props.setTextFilter(e.target.value)
    }
    onSortChange = e=>{
       if(e.target.value === 'date') {
           this.props.sortByDate()
       }else if (e.target.value === 'amount') {
           this.props.sortByAmount()
    }}

    render(){
        return ( /* props degen props.filters,value={} di we want to set whatever the current search is */
            /* How do we dispatch from component otpravit input value to the store,How do we do smth anytime the text input gets changed by the user */
            /* onChange handler takes a function,every single time input changes the function fires,e.target degen input */
            /* We need to change redux store,we need to use dispatch in order to update store*/
            /*Are keystrokes result in a change to the input because input value is always going to be text value on the store  */
            /* When we connect a component to redux we get some other information passed into the component,dev tool-dan koryge bolad naidesh componenty <ExpenseListFilters/> i tam krome filters u nas eshe dostyp k dispatch est==>props.dispatch*/
            /* Every single time we take a keystroke,input baskan sayin,props.dispatch(setTextFilter('e')) fires */
            /*For example we typed 'e' bukvy its going to get added onto the end,so now new state will be with an 'E' on the end,then we hit 'a' its going to add it onto the end  */
            /* kazhdyi raz input value changes onChange my otpravlyaem dispatchom setTextFilter(i bukvu)  */
            /*We have our very 1st component that is not just reading from the redux store,but its writing to the store  */
            /* value={props.filters.text} degen reading off of the store,setTextFilter(e.target.value) degen writing to the store */
            /* <input value="wa"> === dispatch(setTextFilter('wa')) sna4ala dispatch delaet,reducer zapisivaet new value,v vnizu c pomoshu connect() berem eti dannyi, i s stora 4itaem value={props.filters.text} */
            /* zatem /* esli v store proishodit izmenenie,za connectinny componenta <ExpensesList/> zapuskaet connect() i vnutri <ExpensesList/> re-renderitsya obnovlennie stori */
            /*As we change the select it actually changes the store,as the store changes,we're going to make sure to set the value  */
            /*When we set up our form inputs like our text inputs our select dropdowns and we use value onChange we're creating a controlled input.Input where the value is controlled by js  */
                <div> 
                    <input 
                     type="text" 
                     value={this.props.filters.text} 
                     onChange={this.onTextChange}/>
                    <select 
                    value={this.props.filters.sortBy} 
                    onChange={this.onSortChange}>
                        <option value="date">Date</option>
                        <option value="amount">Amount</option>
                    </select>
                <DateRangePicker
                    startDate ={this.props.filters.startDate}
                    endDate ={this.props.filters.endDate}
                    onDatesChange= {this.onDatesChange}
                    focusedInput= {this.state.calenderFocused}
                    onFocusChange= {this.onFocusChange}
                    showClearDates ={true}
                    numberOfMonths={1}
                    isOutsideRange={()=> false}
                 />     
                </div>
            )/*When someone picks option we figure out what value they picked.We'll use value="amount" behind the sceenes  */
            /* we have to set the value and we have to handle changes */
            /*Sort their data out.We also want to give them a way to sort by amount,we will be setting up a select dropdown so they can pick the way they want to sort and view their expenses  */
            /* The goal inside of here is to allow us to change those filters */
    }
}


const mapStateToProps = state =>({ /* on this object what we want off of the store? just filters we dont need expenses*/
    filters: state.filters
}) /*state degen {expenses: [{},{}], filters: {text:'asd'}} want we want off of the store?Zdez opredelim */
     

const mapDispatchToProps = dispatch =>({
    setTextFilter: (text)=> dispatch(setTextFilter(text)),
    sortByDate: ()=> dispatch(sortByDate()),
    sortByAmount: ()=> dispatch(sortByAmount()),
    setStartDate: (startDate)=> dispatch(setStartDate(startDate)),
    setEndDate: (endDate)=> dispatch(setEndDate(endDate)),
})

/* this is connected version of  <ExpenseListFilters />, */
export default connect(mapStateToProps,mapDispatchToProps) (ExpenseListFilters) //ExpenseListFilters degen <ExpenseListFilters/>
//mapStateToProps degen v functione budem imeet dostyp k redux store