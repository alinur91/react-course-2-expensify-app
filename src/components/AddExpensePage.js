import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {addExpense} from '../actions/expenses'

export class AddExpensePage extends React.Component{
  onSubmit=expense=>{ /* it takes the data and passed it through */
    // props.dispatch(addExpense(expense))
     this.props.addExpense(expense)
     this.props.history.push('/') /*Components we render inside of react-router they get a bunch of special props==> props.history.push  */
{/* dashboard pagege redirect / ,this is going to switch me over as if i clicked the link,which means we're not going through full page refresh  */}
   }
   render(){
     return (
        /* onSubmit ishindegi function gets called when the form gets submitted,with valid data
 and we get that data back,expense {description:'asd',amount:'asd'},4ere props peredaem onSumbut*/
 /*this.props degen 4erez instance props onSubmit peredali <ExpenseForm  onSumbit(expense=>{console.log(expense)})/>onSubmit degen callback koi 4erep props peredaem i sdes vizivaem a v propse v functione polu4im expense,shtoby store.dispatch(addExpense(expense)) sdelat 4toby za renderit etot expense  */
 /* expense page needs to be able to dispatch the given action to the redux store */
    <div> 
      <h1>Add Expense</h1>
        <ExpenseForm onSubmit={this.onSubmit}/> {/*push degen kak-budto link basty birey  */}
    </div>
     )
   }
}

const mapDispatchToProps = (dispatch) =>({
    addExpense: (expense)=> dispatch(addExpense(expense))
  })
//the goal is to return an object
  /* on here we define various props and these props are going to call dispatch */
    
/*When we're adding and edditing we can dispatch different stuff  */

/* 1st function is mapsStatesToProp=mynda undefined */
export default connect(undefined,mapDispatchToProps) (AddExpensePage);  
/* za connectili AddExpensePage componenty, and now we have access to props.dispatch*/
/* we do need to pass that component into the 2nd one */