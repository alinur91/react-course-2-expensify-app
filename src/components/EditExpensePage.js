import React from 'react';
import {connect} from 'react-redux'; /* hotim podklu4tisya k redux store
step2 --> it's going to be to search the expenses array for expense with and id that matches this one {props.match.params.id} */
import {editExpense,removeExpense} from '../actions/expenses';
import ExpenseForm from './ExpenseForm';
 
export class EditExpensePage extends React.Component {
  onSubmit= (expense) =>{ /* v <ExpenseForme/> vizivaem onSubmit() a zdes prinimaem */
//     this.props.onSubmit({ /* zdes peredaem object,v AddExpensePage prinimaem object dispatchim shtoby renderilsya expense,i redirecr props.history.push('/') */
//     description: this.state.description,
//     amount: parseFloat(this.state.amount) * 100, /*this.state.note degen 12.12 nado preobrazavat  */
//     note: this.state.note,
//     createdAt: this.state.createdAt.valueOf(), /*this.state.createdAt degen moment object  */
   
// }) /*this.props.expense.id degen minay goi ===> expense: state.expenses.find(expense=> expense.id === props.match.params.id)  */
    this.props.editExpense(this.props.expense.id,expense) /* zdez prosto vizivaem editExpense(expense),v mapDispatchToProps-e dispatchom peredaem*/
    this.props.history.push('/') /*Components we render inside of react-router they get a bunch of special props==> props.history.push  */
  }
  onRemove= (expense) =>{ //this.props.expense.id degen mapStateToProps functionda retun bolady {expense: {}}
    this.props.removeExpense({id: this.props.expense.id})
    this.props.history.push('/') /*Components we render inside of react-router they get a bunch of special props==> props.history.push  */
  }
  
  render(){
    console.log(this.props.expense)
    return ( /* onSubmit= {this.onSubmit} degen v <ExpenseForm/>e vizivaem onSubmit() */
      <div>
        <ExpenseForm
          expense ={this.props.expense}
         onSubmit= {this.onSubmit}
         />
         <button onClick={this.onRemove} >remove</button>
      </div>
    )
  }
}

/*The difference between the AddExpensePage and EditExpensePage is <AddExpensePage/> is fine with those empty defaults and EditExpensePage isn't fine */
/* It wants to be able to overwrite the current values for that expense  */
 /* HIGher order component pen EditExpensePage componenta both have access to props */
    // console.log(props) /* <Router path="/create" component={AddExpensePage}> passing stuff in and that stuff can be useful as we start building more real world app */
/*Any time we use a component inside of a route <BrowserRouter>  we get access to special information(like location,params,querystring)  */
/* kogda zahodim v etot put' <Route path="/edit/:id" component={EditExpensePage} /> my izvlekaem id with the help of {props.match.params.id} */
    
/*  props.history.push('/')  degen redirect to main page */

const mapStateToProps = (state,props) =>{ /* state degen [{descr:'water',amount:23},{}] */
/* props degen <Route path="/edit/:id" component={EditExpensePage} /> osy goi path */
/* what are we searching for? {props.match.params.id} osy id-di expense izdeb zhurmiz */
/*How do we access it? We have access to the props right here.EditExpensePage-tin propsi */
/*We can take some of the current props that were passed into the higher order component and we can use them to calculate the props that we want to add on  */
/* react router it renders our higher order component,higher order component passes the props through and it also allows us to add on some new ones  */
  return { /* return true bolsa callback ishinde 1-yi item tabady,false bolsa eshkaisysi */
    expense: state.expenses.find(expense=> expense.id === props.match.params.id)
  }/* tobedegi <EditExpensePage/> componenta 4erez props access boladi props.expense dep */
}

const mapDispatchToProps = (dispatch) =>(
  { /* we get the data in and pass it through expense v samomo componente vizvali this.props.editExpense(expense) */
    editExpense: (id,expense)=> dispatch(editExpense(id,expense)), //editExpense vnutri etogo functiona dispatchim
    removeExpense: (data)=> dispatch(removeExpense(data))
  }
)

/* nado za connectit etot <EditExpensePage/> componenty k redux store.
Potomy shto we want to be able to get the entire expense object*/
  export default connect(mapStateToProps,mapDispatchToProps) (EditExpensePage);