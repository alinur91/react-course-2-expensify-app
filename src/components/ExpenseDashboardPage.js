import React from 'react';
import ExpensesList from './ExpensesList' /*ExpensesList degen export default  connect(mapStateToProps) (ExpensesList)  */
import ExpenseListFilters from './ExpenseListFilters'

const ExpenseDashboardPage = () =>( /* <ExpensesList /> vernet <h1>Expense List</h1> <h3>Water bill.</h3> <p>123 - 455</p>*/ 
/*<ExpenseListFilters i <ExpensesList /> componety budut imet dostyp k redux store,4erez connect/>  */
    <div> 
      <ExpenseListFilters />
      <ExpensesList />
    </div>
  );
  
export default ExpenseDashboardPage;