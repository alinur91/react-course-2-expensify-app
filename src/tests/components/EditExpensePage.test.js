import React from 'react'
import { shallow } from "enzyme"
import expenses from '../fixtures/expenses'
import { EditExpensePage } from "../../components/EditExpensePage"

let editExpense,removeExpense,history, wrapper;
//we use beforeEach to define variable up above
beforeEach(()=>{
    editExpense = jest.fn() /* spy */
    removeExpense = jest.fn() /* spy */
    history = {push: jest.fn()} //this.props.history.push('/')
    //render that component EditExpensePage
wrapper = shallow(<EditExpensePage  
editExpense={editExpense} 
removeExpense={removeExpense}
history = {history}
expense = {expenses[0]}
/>)
})

test('should  render EditExpensePage', ()=>{
    expect(wrapper).toMatchSnapshot()
/* We have all of the data correctly getting passed down into ExpenseForm */
})

test('should  handle editExpense', ()=>{
// To make sure that EditExpense gets called, i need to do is trigger onSubmit on expense form and we can use spy to make sure the correct data was passed and push and editexpense
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0])
    /* what should happen to those spyies? */
    expect(history.push).toHaveBeenLastCalledWith('/')
    /* what editExpense called with */
expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id,expenses[0])

})

test('should  handle removeExpense', ()=>{
    // To make sure that EditExpense gets called, i need to do is trigger onSubmit on expense form and we can use spy to make sure the correct data was passed and push and editexpense
    wrapper.find('button').simulate('click')
    expect(history.push).toHaveBeenLastCalledWith('/')
expect(removeExpense).toHaveBeenLastCalledWith({id: expenses[0].id})

})