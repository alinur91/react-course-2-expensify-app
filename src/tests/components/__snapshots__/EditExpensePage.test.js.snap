// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should  render EditExpensePage 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 1120,
        "createdAt": 0,
        "description": "xbox",
        "id": "123",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    remove
  </button>
</div>
`;
